### PR TITLE
(PUP-11428) Reload ssl context for each agent run

### PR DESCRIPTION
**Pull request by joshcooper**
* Agents never reloaded their ssl state, which prevented `crl_refresh_interval` from working as intended when `Puppet[:onetime]` was false. Now we reload ssl state prior to each run.
* Clear `default_*_ssl_context` for each run.
* Print CRL information at debug level along with CA and client certs.
* Daemonized agents never logged that they were exiting if they failed to get a cert.
* Added integration tests for bootstrapping SSL.

### CODE REVIEW

**Comment by joshcooper** [comment]
jenkins please test this with servertests

**Comment by joshcooper** [comment]
jenkins please test this with servertests

File: lib/puppet/agent.rb [hunk 1]
```diff
@@ -84,6 +94,8 @@ def run(client_options = {})
           rescue StandardError => detail
             Puppet.log_exception(detail, _("Could not run %{client_class}: %{detail}") % { client_class: client_class, detail: detail })
             nil
+          ensure
+            Puppet.runtime[:http].close
```
Comment (cthorn42) [inline lib/puppet/agent.rb:98]: Trying to wrap my head around why this was added, is it to reset the settings for `default_system_ssl_context` and `default_system_ssl_context`?

Comment (joshcooper) [reply lib/puppet/agent.rb:98]: Yeah, it's to make sure that state is cleared across agent runs.

**Review by cthorn42 — APPROVED** [review]
Looks good, the learning session helped a lot with wrapping my head around this change. 👍

**Pull request merged**